notify/src/subscription/single.rs

```diff
@@ -143,162 +167,386 @@ impl Subscription for VirtualChainChangedSubscription {
+pub enum UtxosChangedState {
+    /// Inactive
+    #[default]
+    None,
+
+    /// Active on a set of selected addresses
+    Selected,
+
+    /// Active on all addresses
+    All,
+}
+
+impl UtxosChangedState {
+    pub fn active(&self) -> bool {
+        match self {
+            UtxosChangedState::None => false,
+            UtxosChangedState::Selected | UtxosChangedState::All => true,
+        }
+    }
+}
+
+impl Display for UtxosChangedState {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            UtxosChangedState::None => write!(f, "none"),
+            UtxosChangedState::Selected => write!(f, "selected"),
+            UtxosChangedState::All => write!(f, "all"),
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct UtxosChangedSubscriptionData {
+    /// State of the subscription
+    ///
+    /// Can be mutated without affecting neither equality nor hash of the struct
+    state: UtxosChangedState,
+
+    /// Address indexes in `SubscriptionContext`
+    ///
+    /// Can be mutated without affecting neither equality nor hash of the struct
+    indexes: Indexes,
+}
+
+impl UtxosChangedSubscriptionData {
+    fn with_capacity(state: UtxosChangedState, capacity: usize) -> Self {
+        let indexes = Indexes::with_capacity(capacity);
+        Self { state, indexes }
+    }
+
+    #[inline(always)]
+    pub fn update_state(&mut self, new_state: UtxosChangedState) {
+        self.state = new_state;
+    }
+
+    pub fn contains(&self, spk: &ScriptPublicKey, context: &SubscriptionContext) -> bool {
+        context.address_tracker.contains(&self.indexes, spk)
+    }
+
+    pub fn len(&self) -> usize {
+        self.indexes.len()
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.indexes.is_empty()
+    }
+
+    pub fn capacity(&self) -> usize {
+        self.indexes.capacity()
+    }
+
+    pub fn iter(&self) -> hash_set::Iter<'_, Index> {
+        self.indexes.iter()
+    }
+
+    pub fn contains_address(&self, address: &Address, context: &SubscriptionContext) -> bool {
+        context.address_tracker.contains_address(&self.indexes, address)
+    }
+
+    pub fn to_addresses(&self, prefix: Prefix, context: &SubscriptionContext) -> Vec<Address> {
+        self.indexes.iter().filter_map(|index| context.address_tracker.get_address_at_index(*index, prefix)).collect_vec()
+    }
+
+    pub fn register(&mut self, addresses: Vec<Address>, context: &SubscriptionContext) -> Result<Vec<Address>> {
+        Ok(context.address_tracker.register(&mut self.indexes, addresses)?)
+    }
+
+    pub fn unregister(&mut self, addresses: Vec<Address>, context: &SubscriptionContext) -> Vec<Address> {
+        context.address_tracker.unregister(&mut self.indexes, addresses)
+    }
+
+    pub fn unregister_indexes(&mut self, context: &SubscriptionContext) -> Vec<Address> {
+        // TODO: consider using a provided prefix
+        let removed = self.to_addresses(Prefix::Mainnet, context);
```

Comment:
`Prefix::Mainnet`: the prefix should be dynamic and based on actual network. After fixing, the TODO above should be removed